pkg/provisioningv2/rke2/planner/encryptionkeyrotation.go

```diff
@@ -96,192 +158,374 @@ func shouldRestartEncryptionKeyRotation(cp *rkev1.RKEControlPlane) bool {
 
 // hasValidRotateEncryptionKeyStatus verifies that the control plane encryption key status is an expected value
 func hasValidRotateEncryptionKeyStatus(cp *rkev1.RKEControlPlane) bool {
-	return cp.Status.RotateEncryptionKeysPhase == rkev1.RotateEncryptionKeysPhaseStart ||
-		cp.Status.RotateEncryptionKeysPhase == rkev1.RotateEncryptionKeysPhaseRestartLeader ||
-		cp.Status.RotateEncryptionKeysPhase == rkev1.RotateEncryptionKeysPhaseVerifyLeaderStatus ||
-		cp.Status.RotateEncryptionKeysPhase == rkev1.RotateEncryptionKeysPhaseRestartFollowers ||
+	return cp.Status.RotateEncryptionKeysPhase == rkev1.RotateEncryptionKeysPhaseRestartNodes ||
 		cp.Status.RotateEncryptionKeysPhase == rkev1.RotateEncryptionKeysPhaseApplyLeader ||
 		cp.Status.RotateEncryptionKeysPhase == rkev1.RotateEncryptionKeysPhaseDone
 }
 
+// restartEncryptionKeyRotation restarts encryption key rotation if it is required, by resetting all related control
+// plane statuses.
 func (p *Planner) restartEncryptionKeyRotation(cp *rkev1.RKEControlPlane) error {
-	err := p.encryptionKeyRotationUpdateControlPlanePhase(cp, rkev1.RotateEncryptionKeysPhaseStart)
+	logrus.Infof("rkecluster %s/%s: restarting encryption key rotation, current phase: [%s]", cp.Namespace, cp.Spec.ClusterName, cp.Status.RotateEncryptionKeysPhase)
+
+	cp = cp.DeepCopy()
+	rke2.Ready.Unknown(cp)
+	rke2.Ready.Message(cp, "encryption key rotation: restarting nodes")
+	rke2.Ready.Reason(cp, rke2.RotatingEncryptionKeysReason)
+	if cp.Status.RotateEncryptionKeys == nil {
+		cp.Status.RotateEncryptionKeys = &rkev1.RotateEncryptionKeysStatus{}
+	}
+	cp.Status.RotateEncryptionKeys.LastRestart = time.Now().Format(time.RFC3339)
+	cp.Status.RotateEncryptionKeysPhase = rkev1.RotateEncryptionKeysPhaseRestartNodes
+	_, err := p.rkeControlPlanes.UpdateStatus(cp)
 	if err != nil {
 		return err
 	}
 	return generic.ErrSkip
 }
 
-func (p *Planner) encryptionKeyRotationStart(cp *rkev1.RKEControlPlane, leader *planEntry) error {
-	return p.encryptionKeyRotationGetLeaderStatus(cp, leader, rkev1.RotateEncryptionKeysPhaseRestartLeader)
-}
+// encryptionKeyRotationElectLeader returns the current leader if it is valid, otherwise it selects the first suitable
+// control plane it can find.
+func (p *Planner) encryptionKeyRotationElectLeader(cp *rkev1.RKEControlPlane, clusterPlan *plan.Plan) (leader *planEntry, init *planEntry, err error) {
```

Comment:
I won't block merging on this, but I really don't like this.

In this case, we don't gain anything from having named return values in this case.